File: packages/python/src/alumnium/server/api_models.py
from typing import Any, Literal

from pydantic import BaseModel, Field


# Base versioned model
class VersionedModel(BaseModel):
    api_version: str = Field(default="v1", description="API version")


class SessionRequest(VersionedModel):
    platform: Literal["chromium", "uiautomator2", "xcuitest"]
    provider: str
    name: str | None = None
    tools: list[dict[str, Any]]
    planner: bool = True


class SessionResponse(VersionedModel):
    session_id: str


class PlanRequest(VersionedModel):
    goal: str
    accessibility_tree: str
    url: str | None = None
    title: str | None = None


class PlanResponse(VersionedModel):
    explanation: str
    steps: list[str]


class StepRequest(VersionedModel):
    goal: str
    step: str
    accessibility_tree: str


class StepResponse(VersionedModel):
    explanation: str
    actions: list[dict[str, Any]]


class StatementRequest(VersionedModel):
    statement: str
    accessibility_tree: str
    url: str | None = None
    title: str | None = None
    screenshot: str | None = None  # base64 encoded image


class StatementResponse(VersionedModel):
    result: str | list[str]
    explanation: str


class AreaRequest(VersionedModel):
    description: str
    accessibility_tree: str


class AreaResponse(VersionedModel):
    id: int
    explanation: str


class FindRequest(VersionedModel):
    description: str
    accessibility_tree: str


class FindResponse(VersionedModel):
    elements: list[dict[str, int | str]]


class AddExampleRequest(VersionedModel):
    goal: str
    actions: list[str]


class AddExampleResponse(VersionedModel):
    success: bool
    message: str


class ClearExamplesResponse(VersionedModel):
    success: bool
    message: str


class CacheResponse(VersionedModel):
    success: bool
    message: str


class ChangeState(VersionedModel):
    accessibility_tree: str
    url: str


class ChangesRequest(VersionedModel):
    before: ChangeState
    after: ChangeState


class ChangesResponse(VersionedModel):
    result: str


class ErrorResponse(VersionedModel):
    error: str
    detail: str | None = None

File: packages/typescript/src/clients/ApiModels.ts
/**
 * TypeScript interfaces for API request and response models.
 *
 * IMPORTANT: These interfaces must be kept in sync with the Python API models
 * defined in packages/python/src/alumnium/server/api_models.py
 */

export interface SessionRequest {
  platform: "chromium" | "uiautomator2" | "xcuitest";
  provider: string;
  name?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  tools: { [key: string]: any }[];
  planner?: boolean;
}

export interface SessionResponse {
  session_id: string;
}

export interface PlanRequest {
  goal: string;
  accessibility_tree: string;
  url?: string;
  title?: string;
}

export interface PlanResponse {
  explanation: string;
  steps: string[];
}

export interface StepRequest {
  goal: string;
  step: string;
  accessibility_tree: string;
}

export interface StepResponse {
  explanation: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  actions: { [key: string]: any }[];
}

export interface StatementRequest {
  statement: string;
  accessibility_tree: string;
  url?: string;
  title?: string;
  screenshot?: string | null;
}

export interface StatementResponse {
  result: string | string[];
  explanation: string;
}

export interface AreaRequest {
  description: string;
  accessibility_tree: string;
}

export interface AreaResponse {
  id: number;
  explanation: string;
}

export interface FindRequest {
  description: string;
  accessibility_tree: string;
}

export interface FindResponse {
  elements: { [key: string]: number | string }[];
}

export interface AddExampleRequest {
  goal: string;
  actions: string[];
}

export interface StatsResponse {
  [key: string]: {
    [key: string]: number;
  };
}
